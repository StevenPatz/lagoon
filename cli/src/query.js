// @flow

import path from 'path';
import os from 'os';
import url from 'url';
import R from 'ramda';
import { fileExists, readFile } from './util/fs';
import request from './util/request';
import { printErrors } from './printErrors';

type QLQueryArgs = {
  cerr: typeof console.error,
  endpoint?: string,
  query: string,
  variables?: Object,
  headers?: Object,
  pretty?: boolean,
};

export async function runGQLQuery(args: QLQueryArgs): Object {
  const {
    cerr,
    endpoint = process.env.API_URL || 'https://api.amazee.io/graphql',
    query,
    variables,
    headers: customHeaders = {},
    pretty = false,
  } = args;

  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...customHeaders,
  };

  if (!headers.Authorization) {
    const tokenFile = path.join(os.homedir(), '.ioauth');
    const tokenFileExists = await fileExists(tokenFile);

    if (tokenFileExists) {
      const tokenBuffer = await readFile(tokenFile);
      const token = tokenBuffer.toString().replace(/(\r\n|\n|\r)/gm, '');
      headers.Authorization = `Bearer ${token}`;
    }
  }

  const {
    hostname,
    path: pathname,
    port: urlPort,
    protocol = 'https:',
  } = url.parse(endpoint);

  if (hostname == null) {
    throw new Error('Hostname required');
  }

  const body = JSON.stringify(
    {
      query,
      variables,
    },
    null,
    pretty ? 2 : 0,
  );

  const protocolPorts = {
    'https:': 443,
    'http:': 80,
  };

  const options = {
    hostname,
    path: pathname,
    port:
      urlPort === null
        ? R.propOr(443, protocol)(protocolPorts)
        : Number(urlPort),
    method: 'POST',
    headers,
    body,
    rejectUnauthorized: false,
  };

  try {
    return await request(options);
  } catch (err) {
    const error = R.ifElse(
      R.propEq('message', 'socket hang up'),
      // Print a nicer error message for socket hangups
      R.always('Could not connect to API.'),
      // If not a socket hang up, return the error message
      R.prop('message'),
    )(err);
    printErrors(cerr, error);
    process.exit(1);
  }
}

export default runGQLQuery;
