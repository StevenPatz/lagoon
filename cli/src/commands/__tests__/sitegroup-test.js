// @flow

import { runGQLQuery } from '../../query';
import { sitegroupInfo } from '../sitegroup';

jest.mock('../../query');

const _mock = (mockFn: any): JestMockFn => mockFn;

const mockErrorResponse = {
  errors: [{ message: 'something something error' }],
};

describe('siteGroupInfo', () => {
  const mockResponse1 = {
    data: {
      siteGroupByName: {
        gitUrl: 'sitegroup.git',
        siteGroupName: 'mysitegroup',
        slack: {
          webhook: 'https://slack-webhook.something',
          channel: 'myslack',
        },
        client: {
          clientName: 'me',
        },
        sites: [
          {
            siteName: 'site1',
            siteBranch: 'dev',
          },
          {
            siteName: 'site1',
            siteBranch: 'prod',
          },
        ],
      },
    },
  };

  it('should display error, if GraphQL sends error messages', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve(mockErrorResponse),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await sitegroupInfo({
      sitegroup: 'some_sitegroup',
      clog,
      cerr,
    });

    expect(code).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it('should show error on missing sitegroup', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() => Promise.resolve({}));

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await sitegroupInfo({
      sitegroup: 'not_existing',
      clog,
      cerr,
    });

    expect(code).toBe(1);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should list found information for given sitegroup', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve(mockResponse1),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await sitegroupInfo({
      sitegroup: 'mysitegroup',
      clog,
      cerr,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });
});
