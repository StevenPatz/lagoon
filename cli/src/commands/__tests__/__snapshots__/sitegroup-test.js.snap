// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`siteGroupInfo should display error, if GraphQL sends error messages 1`] = `
Array [
  Array [
    "[31mOops! The amazee.io API returned an error:[39m",
  ],
  Array [
    "[31m{[39m
[31m  \\"message\\": \\"something something error\\"[39m
[31m}[39m",
  ],
]
`;

exports[`siteGroupInfo should list found information for given sitegroup 1`] = `
Array [
  Array [
    "Details for 'mysitegroup':",
  ],
  Array [
    "╔════════════════╤════════════════════════════════════════════╗
║ Sitegroup Name │ mysitegroup                                ║
╟────────────────┼────────────────────────────────────────────╢
║ Git Url        │ sitegroup.git                              ║
╟────────────────┼────────────────────────────────────────────╢
║ Slack          │ myslack -> https://slack-webhook.something ║
╟────────────────┼────────────────────────────────────────────╢
║ Sites          │ site1:dev, site1:prod                      ║
╚════════════════╧════════════════════════════════════════════╝
",
  ],
]
`;

exports[`siteGroupInfo should show error on missing sitegroup 1`] = `
Array [
  Array [
    "[31mNo sitegroup 'not_existing' found[39m",
  ],
]
`;
